### ISSUE TITLE
Idea: `compose` could accept an array of functions

### ISSUE THREAD

**Issue by tomekwi** [tjmehta/101#53]
One use case:

```js
[ {a: true, b: false}
, {a: false, b: false}
, {a: true, b: true}
].filter(compose(and
  , [ pluck("a")
    , pluck("b")
    ]
  ));
// » [{a: true, b: true}]
```

**Comment by stoeffel** [comment]
I think this use case is to specific. But I would like to have a `composeN` function where you could pass more then two functions. So you can create a new function like this: `composeN(f, g, h, i) === f(g(h(i)))`. This would be more generic IMHO.

**Comment by tomekwi** [comment]
@stoeffel isn't it already so? The problem with both: you can only use functions which take one argument.

I meant a generic way to overcome this:
```js
compose(f, [g, h]) == f(g, h);

compose(f, g, [h, i]) == f(g(h, i));

compose(f, g, [h, i], j) == f(g(h(j), i(j)));

compose(f, [g, compose(h, i)]) == f(g, h(i));
```

**Comment by stoeffel** [comment]
> isn't it already so?

Nope.

Yes, you can only use functions which take one argument, but this is how I expect a compose function to work. Checkout other libs (i.e. lodash). Maybe a new function for your use case would be better. How about `branch`?

**Comment by tomekwi** [comment]
So again it looks the discussion between overloading and separate functions. I'd go for overloading both your proposal @stoeffel and the `branch` functionality into `compose`.

Perhaps @tjmehta can chime in?

**Comment by stoeffel** [comment]
In my opinion this should be two different functions, since they do two different things. One composes a function and one converge multiple functions to one. But the last word has @tjmehta :smile_cat:

**Comment by tjmehta** [comment]
Quick thoughts for now, I definitely see the scenario shown by @tomekwi's example. 
I don't know if overloading the compose function is the right solution here. Reading the
code I would not necessarily understand what behavior the array is producing. 
I also think that functions other than compose may be benefited with some transform argument behavior.

Just throwing this out there for discussion: 
You could used [tjmehta/function-proxy](http://github.com/tjmehta/function-proxy) to transform arguments.
But once you have more that a few functions wrapping one another the code can be hard to read..

**Comment by stoeffel** [comment]
How about `converge`

```js
/**
 * Converge a new function from the given functions
 * as arguments into the first function
 * @function module:101/converge
 * @param {function} func
 * @param {Array} funcs
 * @return {function} 
 */
module.exports.converge = function(func, argFuncs) {
  return converged = function() {
     var args = Array.prototyp.slice.call(arguments);
     var convergedArgs = argFuncs.map(function(f) {
       return f.apply(null, args);
     });
    return func.apply(null, convergedArgs);
  };
};
```

**Comment by tomekwi** [comment]
Cool @stoeffel – does that mean the following?

```
compose(f, converge(g, h))  ~=  f(g, h);

compose(f, g, converge(h, i))  ~=  f(g(h, i));

compose(f, converge(g, h), i)  ~=  f(g(i), h(i));

compose(f, converge(g, compose(h, i)))  ~=  f(g, h(i));
```

**Comment by tomekwi** [comment]
– or rather as per #56 

```
[f, converge(g, h)].reduce(compose)  ~=  f(g, h);

[f, g, converge(h, i)].reduce(compose)  ~=  f(g(h, i));

[f, converge(g, h), i].reduce(compose)  ~=  f(g(i), h(i));

[f, converge(g, compose(h, i))].reduce(compose)  ~=  f(g, h(i));
```

**Comment by stoeffel** [comment]
@tomekwi `converge` would more be like this
```js
var add = function(x, y) { return x + y };
var sub = function(x, y) { return x - y };
var mul = function(x, y) { return x * y };
converge(mul, [add, sub]))(2, 3);  === mul(add(2, 3), sub(2, 3)); // => -5
```

But you could use it in a `reduce(compose)`.
```js
[f, g, converge(h, [i, j])].reduce(compose)  ~=  f(g(h(i, j)));
```

**Comment by stoeffel** [comment]
I can create a PR for `converge` if you guys like it :smile:

**Comment by tomekwi** [comment]
I love the idea @stoeffel! The name is great as well :)

I'm just wondering about the API – if it doesn't overlap with `compose` a little, the way it now is? After all

```
compose(a, b)  ~=  converge(a, [b]);
```

isn't it so?

But then again, I have no other idea. `compose(mul, converge(add, sub))` is in no way doable, is it?

**Comment by stoeffel** [comment]
> if it doesn't overlap with compose a little, the way it now is? 

Yes, that's true. But I think it's not a problem.

> I have no other idea. compose(mul, converge(add, sub)) is in no way doable, is it?

No, it isn't

> I love the idea @stoeffel! The name is great as well :)

Cool. I will create a PR and we can all discuss more there.

**Comment by tjmehta** [comment]
I like the idea of adding converge to 101, that's exactly what @tomekwi want. Thanks @stoeffel!

**Comment by tjmehta** [comment]
Published converge in 101@v0.10.0

**Comment by tomekwi** [comment]
:+1:
